catalog/models.py

```diff
@@ -127,7 +128,13 @@ def tagged(self, tags: 'TagQuerySet'):
         # See docs for details:
         # https://www.postgresql.org/docs/10/static/sql-select.html#SQL-DISTINCT
         # https://docs.djangoproject.com/en/2.1/ref/models/querysets/#django.db.models.query.QuerySet.distinct
-        return self.filter(tags__in=tags).distinct()
+        # Q expressions doesn't work for this case.
+        # That's why we use qs.filter reduce.
+        return reduce(
+            lambda accum, group: accum.filter(tags__in=group),
+            (group for _, group in groupby(tags, lambda t: t.group)),
```

Comment:
`groupby` requires ordered values by a grouping key as the first argument, so we should order it by `t.group`

I'd prefer to use `dict` instead, because it is explicit, reliable and it will bring better performance than `groupby`

So either fix `tags` order, or use `dict`

Reply:
@artemiy312 , what do you mean by using dict? Provide plz one string of code or link on the doc

Reply:
@duker33 
```python
grouped = defaultdict(list)
for t in tags:
    grouped[t.group].append(t)
```